### ISSUE TITLE
SageTV sorts numbers by first digit instead of all digits

### ISSUE THREAD

**Issue by 49studebaker** [google/sagetv#126]
On the "Videos" screen, SageTV sorts number by the first digit instead of the entire number.  

Example: 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 3, 4, 5, 6, 7, 8, 9.

**Comment by CraziFuzzy** [comment]
I was thinking about trying to write a more comprehensive title comparator for use in sage.  Not just to handle 'natural' number in string sorting, but to detect and deal with roman numerals as well.  Actually parsing the roman numerals into integers, and from there comparing them numerically.  This would help some of those really annoying cases such as:

Men in Black
Men in Black II
Men in Black 3

**Comment by Narflex** [comment]
If somebody does this, you basically need to deal with it in the Database
API class. There are a lot of comparators in there, but probably some way
to make it more universal.

Jeff Kardatzke

**Comment by CraziFuzzy** [comment]
Yeah, started browsing through the API code a bit last night (as well as browsing various comparator examples online).  I did notice that the MediaFile object does have comparators, but it seems to only be either AiringID or FilePath - nothing based on title (would be great to have a special comparator in the object itself that compared Title, Episode, etc all together to give a good sorting).

**Comment by Narflex** [comment]
The comparators in the child classes of DBObject (MediaFile in this case)
are used for maintaining the alternate lookup indices in the database
(Wizard). :-)

Jeff Kardatzke

**Comment by CraziFuzzy** [comment]
Ahhh.. gottcha.  I was hoping the API would hand off comparing to an object's comparator if it had one, so an array of MediaFiles would get sorted as MediaFiles, instead of essentially running the .toString through a string comparator at the API level.  So, yeah, I'll have to look into the database sort method and just take in the strings then.

**Comment by Narflex** [comment]
You don't want to mess with the comparators in the DBObject classes (just
to be clear) that are used by Wizard (there's other reasons why, but this
would not be one). Changing the ones in the Database API are fine.

Jeff Kardatzke

**Comment by CraziFuzzy** [comment]
So i think I've written a robust enough comparator to deal with this, but haven't looked into he best way to wedge it into the API yet.  Thought I'd give others a chance to look it over and throw some test cases at it to see if I've missed a usage.  As far as i know, the only potential issue is if the title has a word that IS a roman numeral, it's going to treat that as a number.  I can only really see this happening with I, but I'm not positive the best way to handle that.  Probably, since I don't thing there are going to be many items with a roman numeral 1 ("I") in it, I should specifically ignore that word, always treating it as a letter.
In any case, here's the code I've worked out in a test routine:
http://ideone.com/kkyn39

**Comment by enternoescape** [comment]
That code ends up creating a lot of String objects when it's doing the roman numerals. The code very easy to follow in it's current form, but it will probably perform faster on very large collections if you used index and length instead of substring.

My understanding is that as of Java 7, unless the string is interned, it will create a new char array every time you create a substring. When I do sorting, I try not to introduce new objects because every new object is time and it all adds up. Then again, maybe I'm just suggesting some premature optimization.

**Comment by CraziFuzzy** [comment]
Yeah, like I said, the performance was my biggest concern, but I haven't put it into sage to actually test it yet.  In reality, we in our use probably don't need to count Romans higher than about 20, but if making a thing do a thing, if should be able to do all the things.

**Comment by enternoescape** [comment]
The reason I was talking about premature optimization is just the simple fact that out of let's say 1000 movie titles, how many are actually going to have a roman numeral in the title. I did a quick test on my own movie collection and it didn't catch **Back to the Future Part II (1989)** and **Back to the Future Part III (1990)** as containing roman numerals. Also the .toUpperCase at the start is probably going to increase your chances of a false positive since being lowercase could be used as a hint that something is not a roman numeral. This might be more complicated than it looks at first glance.

**Comment by CraziFuzzy** [comment]
I'm pretty sure the existing sort is case insensitive as well.  (Otherwise, the entire capital alphabet would be before the lower case alphabet), so the uppercase transition is happening already.

Regarding false detection of Roman Numerals, the regex seems relatively robust.  It will only detect an actual valid numeral - not just the a selection of MCLXIV in any order.  It also will only work if following the numeral is a non-number or non-letter (so punctuation, whitespace, and end of string are okay).  I think the only current issue (which I just detected this evening) is if a valid roman numeral string exists at the end of another word, as I don't check what precedes the numeral - I'll have to tweak the regex to clean that up I think.

**Comment by enternoescape** [comment]
It's very normal for this kind of sort to be case insensitive. I'm just trying to give you constructive feedback or maybe another way to look at it.

**Comment by Narflex** [comment]
And I don't think it's actually doing something like toUpperCase when it's
doing the case insensitive comparison; it does this on a character by
character basis instead (which it would have to do anyways) when it does
the comparison to avoid having to do any allocations.

On Fri, Jun 24, 2016 at 7:05 PM, Joseph Shuttlesworth <
notifications@github.com> wrote:

> It's very normal for this kind of sort to be case insensitive. I'm just
> trying to give you constructive feedback or maybe another way to look at it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/sagetv/issues/126#issuecomment-228501365>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ANEIDH4-kq3m-05knDOoSZlL0Y2ZluyGks5qPI0AgaJpZM4IJitN>
> .
>



-- 
Jeffrey Kardatzke
jkardatzke@google.com
Google, Inc.
